File: src/models/runtime-snapshot.ts
/**
 * Runtime snapshot model.
 *
 * A snapshot captures the *live* shell runtime of one machine — installed tool
 * versions, `$PATH` resolution order, optional global packages, and the OS —
 * so two machines can be diffed to explain "builds on my laptop, fails on the
 * server". It is a different data source from the static file model the rest of
 * envdoctor works on, and lives beside the pipeline rather than inside it.
 *
 * The envdoctor invariant still holds: **values are never captured.** Only
 * variable *names* are recorded in `envFlagNames`, and names that look secret
 * are dropped entirely.
 */

/** Bumped whenever the snapshot shape changes; `snapshot-diff` refuses tokens it can't read. */
export const SNAPSHOT_SCHEMA = 1;

export interface ToolVersion {
  /** Probe id, e.g. "node". */
  tool: string;
  /** Normalized version string (digits and dots), e.g. "20.11.1". */
  version: string;
  /** PATH directory the tool resolved from, with $HOME collapsed to "~". */
  resolvedFrom: string;
}

export interface GlobalPackage {
  name: string;
  version: string;
}

export interface RuntimeSnapshot {
  schema: number;
  /** Informational only — never diffed. */
  capturedAt: string;
  os: { platform: string; arch: string; release: string };
  /** Present tools only, sorted by tool id. */
  tools: ToolVersion[];
  /** Ordered `$PATH` entries, $HOME collapsed to "~". Order is significant. */
  path: string[];
  /** Ecosystem id ("npm", "pip", …) → package inventory. Empty unless --globals. */
  globals: Record<string, GlobalPackage[]>;
  /** Non-secret env var NAMES only. Never any values. */
  envFlagNames: string[];
}

File: dart/lib/src/models/runtime_snapshot.dart
/// Runtime snapshot model. A snapshot captures the *live* shell runtime of
/// one machine — installed tool versions, `$PATH` resolution order, optional
/// global packages, and the OS. **Values are never captured.** Only variable
/// *names* are recorded in `envFlagNames`, and names that look secret are
/// dropped entirely.
library;

/// Bumped whenever the snapshot shape changes; `snapshot-diff` refuses tokens
/// it can't read.
const int snapshotSchema = 1;

class ToolInfo {
  final String tool;
  final String version;
  final String resolvedFrom;

  ToolInfo({required this.tool, required this.version, required this.resolvedFrom});
}

class GlobalPackage {
  final String name;
  final String version;

  GlobalPackage(this.name, this.version);
}

class OsInfo {
  String platform;
  String arch;
  String release;

  OsInfo({this.platform = '', this.arch = '', this.release = ''});
}

class RuntimeSnapshot {
  int schema;
  String capturedAt;
  OsInfo os;
  List<ToolInfo> tools;
  List<String> path;
  Map<String, List<GlobalPackage>> globals;
  List<String> envFlagNames;

  /// Set when the parsed input has the minimum snapshot shape
  /// (numeric `schema`, array `tools`).
  bool hasValidShape = true;

  RuntimeSnapshot({
    this.schema = snapshotSchema,
    this.capturedAt = '',
    OsInfo? os,
    List<ToolInfo>? tools,
    List<String>? path,
    Map<String, List<GlobalPackage>>? globals,
    List<String>? envFlagNames,
  })  : os = os ?? OsInfo(),
        tools = tools ?? [],
        path = path ?? [],
        globals = globals ?? {},
        envFlagNames = envFlagNames ?? [];
}
